### ISSUE TITLE
model.select_features/F1 problems

### ISSUE THREAD

**Issue by bottledmind** [catboost/catboost#1688]
Problem: model.select_features/F1 problems
catboost version: 0.25.1
Operating System: Ubuntu 20.04
```model.select_features``` does not train the best model.
Catboost F1 and Sci-kit learn F1 differ significantly
model trained with ```model.select_features```  returns empty dict from ```model.best_score_```
```
import pandas as pd
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
```
```
df = pd.read_csv('df.csv').drop('Unnamed: 0',axis=1)
df.head(1)
```
```
X = df.drop(['Employee','IsDismissed'],axis=1)
y = df.IsDismissed

cat_features = [0, 1, 2, 3, 4, 5, 6, 12, 13, 15, 16, 17, 18, 19, 20, 22]

model = CatBoostClassifier(
    custom_loss=['F1','Accuracy'],
    random_seed=42,
    logging_level='Silent',
    l2_leaf_reg=5,
    cat_features=cat_features,
    class_weights={"True":1,"False":(y).sum()/(~y).sum()},
    eval_metric='F1'
)
```
```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
```
out = model.select_features(
                X=X_train,
                y=y_train,
                eval_set=(X_test,y_test),
                features_for_select='0-22',
                num_features_to_select=10,
                steps=10,
                shap_calc_type=None,
                train_final_model=True,
                plot=True)
```
![image](https://user-images.githubusercontent.com/60937565/117021131-bfd55800-acff-11eb-8902-25f7fa99a247.png)
![image](https://user-images.githubusercontent.com/60937565/117021241-d8457280-acff-11eb-9353-53d0a6c78cc9.png)
![image](https://user-images.githubusercontent.com/60937565/117021340-eeebc980-acff-11eb-9bf5-7eb945bad96b.png)
```
model.best_score_
```
![image](https://user-images.githubusercontent.com/60937565/117021518-180c5a00-ad00-11eb-9942-9a3743487ac7.png)
```
f1_score(y_test,
         pd.Series(model.predict(X_test)).replace(
             {'True':True,
              'False':False})
        )
```
![image](https://user-images.githubusercontent.com/60937565/117021652-340ffb80-ad00-11eb-8a1c-594967506317.png)

**Comment by LyzhinIvan** [comment]
Hi, @bottledmind!

1) What do you mean by "model.select_features does not train the best model"? If you mean that best score of "model-1" and "model-2" is better than "model-final" - it is expected behavior, because they had more available features.
2) F1 score from CatBoost differs from F1 from scikit-learn because of weights. You have specified `class_weights` and CatBoost uses them for metric calculation by default. For correct comparison try to open "F1:use_weights=false" tab at the plot. 
3) Empty `model.best_score_` is a bug, we will fix it.

**Comment by bottledmind** [comment]
@LyzhinIvan
1. Well, I got it now. Still it would be nice to have a param like "chose the best set of params, if max params count is n". I was using this method initially because I had bunch of features that were decreasing the model quality
2. It does not seem sklearn F1 matches any of them
![image](https://user-images.githubusercontent.com/60937565/118030125-8d65d380-b36d-11eb-9732-ee782028ad64.png)
![image](https://user-images.githubusercontent.com/60937565/118030195-a4a4c100-b36d-11eb-8ceb-b0bfbb2a9ef7.png)
![image](https://user-images.githubusercontent.com/60937565/118030290-c3a35300-b36d-11eb-8124-9396fb1c4d6b.png)

**Comment by LyzhinIvan** [comment]
@bottledmind 
1) We are going to add a possibility to eliminate features as long as the loss function improves.
2) I can't reproduce such behavior. I've tried to use your code with Adult dataset and got results on `F1:use_weights=false` exactly same as sklearn `f1_score`. Please check your code, maybe you made some mistake.

**Comment by bottledmind** [comment]
@LyzhinIvan 
Well, I was trying to reproduce it with adult myself, and... I did. For some reason I could not get the dataset with catboost.datasets.adult(the cell was running for quite a while with no result) so I've downloaded it directly from https://archive.ics.uci.edu/ml/machine-learning-databases/adult/. Here's what i have:
```
import pandas as pd
from catboost import CatBoostClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from catboost.datasets import adult
```
```
data = pd.read_csv('/home/some/path/Downloads/adult.data')
```
```
X = data.drop(['39', ' 77516', ' 13', ' 2174', ' 0', ' 40', ' <=50K'],axis=1)
y = data[' <=50K']

cat_features = range(8)

model = CatBoostClassifier(
    custom_loss=['F1','Accuracy'],
    random_seed=42,
    logging_level='Silent',
    l2_leaf_reg=5,
    cat_features=cat_features,
    class_weights={" <=50K":1," >50K":(y==' >50K').sum()/(y!=' >50K').sum()},
    eval_metric='F1'
)
```
```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
```
```
out = model.select_features(
                X=X_train,
                y=y_train,
                eval_set=(X_test,y_test),
                features_for_select='0-7',
                num_features_to_select=4,
                steps=3,
                shap_calc_type=None,
                train_final_model=True,
                plot=True)
```
![image](https://user-images.githubusercontent.com/60937565/119240250-1a234500-bb57-11eb-83ec-810ec879994f.png)
![image](https://user-images.githubusercontent.com/60937565/119240261-26a79d80-bb57-11eb-83bf-9e1fc53cd1da.png)
```
f1_score(y_test.replace(
             {' <=50K':1,
              ' >50K':2}),
         pd.Series(model.predict(X_test)).replace(
             {' <=50K':1,
              ' >50K':2})
        )
```
![image](https://user-images.githubusercontent.com/60937565/119240288-4939b680-bb57-11eb-9629-7c16ae832900.png)
(for some reason f1_score was not working without replacement)
CatBoost was installed with Anaconda:
conda version - 4.10.1
CatBoost version - 0.25.1
sklearn version - 0.24.2

**Comment by LyzhinIvan** [comment]
Hi, @bottledmind! 

Sorry for long response. I think this picture will explain everything to you.
![Screen Shot 2021-10-01 at 6 31 03 PM](https://user-images.githubusercontent.com/6265308/135648781-54d14a2f-dd9b-4cb4-a1d2-44eb593fce72.png)

For adult dataset a small difference exists anyway because of categorical features. If you train without categorical features than there will be no difference with sklearn.

**Comment by LyzhinIvan** [comment]
`model.best_score_` and `model.evals_result_` were fixed in 19bf77bdc8dc593603ffd7aa481f1243d923593b
